mails are of two types-> spam mails, ham mails

-> mail data
-> data pre processing
-> split data in training data and test data
-> and feed the training data to the logistic regression model(best for binary classification problems i.e., 2 classes are there

now when a new model arrives the logictic regression model will be predict if it is a spam mail or ham mail






# Importing the dependencies

In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

numpy is used to create numpy arrays,

and pandas is used to create dataframes(converts csv to more structured data)

train_test_split function splits the data into training and testing data

TfidfVectorizer CONVERTS THE TEXT DATA into more meaningful data(numeric data) or feature vectors

accuracy_store is used to find accuracy of our model when it is tested against the testing data after being trained



# Data collection and pre processing

In [ ]:
#loading the data from csv file to a pandas data frame
raw_mail_data = pd.read_csv("/content/mail_data.csv")

In [ ]:
print(raw_mail_data)

     Category                                            Message
0         ham  Go until jurong point, crazy.. Available only ...
1         ham                      Ok lar... Joking wif u oni...
2        spam  Free entry in 2 a wkly comp to win FA Cup fina...
3         ham  U dun say so early hor... U c already then say...
4         ham  Nah I don't think he goes to usf, he lives aro...
...       ...                                                ...
5567     spam  This is the 2nd time we have tried 2 contact u...
5568      ham               Will ü b going to esplanade fr home?
5569      ham  Pity, * was in mood for that. So...any other s...
5570      ham  The guy did some bitching but I acted like i'd...
5571      ham                         Rofl. Its true to its name

[5572 rows x 2 columns]


In [ ]:
#replacing the null values with a null string
mail_data = raw_mail_data.where((pd.notnull(raw_mail_data)),'')

In [ ]:
#printing the first 5 rows of the dataframe
mail_data.head()

Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...

In [ ]:
#checking the number of ros and column(size of the dataframe)
#O/P -->  (rows, column)
mail_data.shape

(5572, 2)

# Label Encoding

In [ ]:
#label spam mail as 0 and ham mail as 1
mail_data.loc[mail_data['Category'] == 'spam', 'Category',] = 0
mail_data.loc[mail_data['Category'] == 'ham', 'Category',] = 1

In [ ]:
#seperating the data as texts and labels(its like giving x axis value anad y axis value to the model
X = mail_data['Message']

Y = mail_data['Category']

In [ ]:
print(X)

0       Go until jurong point, crazy.. Available only ...
1                           Ok lar... Joking wif u oni...
2       Free entry in 2 a wkly comp to win FA Cup fina...
3       U dun say so early hor... U c already then say...
4       Nah I don't think he goes to usf, he lives aro...
                              ...                        
5567    This is the 2nd time we have tried 2 contact u...
5568                 Will ü b going to esplanade fr home?
5569    Pity, * was in mood for that. So...any other s...
5570    The guy did some bitching but I acted like i'd...
5571                           Rofl. Its true to its name
Name: Message, Length: 5572, dtype: object


In [ ]:
print(Y)

0       1
1       1
2       0
3       1
4       1
       ..
5567    0
5568    1
5569    1
5570    1
5571    1
Name: Category, Length: 5572, dtype: object


# Splitting the data in training and testing data

In [ ]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 3)

x and y will be splited in 2 parts training and testing data 

after splitting x all the y data will  split correspondingly

the X_train, X_test, Y_train, Y_test are arrays

training data is more than testing data
train_test_split(X, Y, test_size = 0.2) will split the data and test size = 0.2 will split the 20% of datA as testing data and remaining 80% of data as training data

each time we use the train_test_split function it';ll split the data in different ways, if we want the data to be splitted in the same way then we mention the random state 3 parameter if we pass random state parater value as 2 it'll slpit it in a different manner

In [ ]:
print(X.shape)
print(X_train.shape)
print(X_test.shape)

(5572,)
(4457,)
(1115,)


here (5572,) secon parameter is empty that means there is only one column as shape gives output as(rows, column)

5572 -> total data(rows)

4457 -> training data

1115 -> testing data

# Feature Extraction

In [ ]:
#transform the text data to feature vectors tha can be used as input to the logistic regression model
#feature vectors are basically numeric values
feature_extraction = TfidfVectorizer(min_df = 1, stop_words='english', lowercase='True')

X_train_features = feature_extraction.fit_transform(X_train)
X_test_features = feature_extraction.transform(X_test)

#convert Y_train and Y_test values as intergers bcz though the value is 0 and 1 but it is in the form of objects/strings and 
#we don't want it to get append so doing this will help(ease) the model to work

Y_train = Y_train.astype('int')
Y_test = Y_test.astype('int')

this TfidfVectorization() looks at the data and look for the words(free, offer, discount etc) in the documents(data set) and if it is repeated several times it will get some values based on how many times the word is being repeated and give the value(importance score)

now the logistic regression model will link all these wordsto get some numerical value and they will be related to the label spam which is zero

and other mails will getsome other score values and will  be linked to ham mails

so the difference betweeen spam and ham mails is identified by going through these values 


PARAMETERS

min_df = 1  ---> 

if the score is more than 1 e need to ignore it as we only have 2 categories 0 & 1 so if the word is repeated only one time we don't care about it bcz it won't be that important for our prediction

this score will be given to all the words individually


stop_words = 'english'--->


did, is, the etc are called stop words bcz they don't hold much value and will be ignored by the logistic regression(LR)


lowercase = 'true'--->



all the letters will be changed to lowercase as it will be easy for processing


X_train_feature will store the numeric value after the x_train data is converted into feature vectors same will be done with x test as they both contain messages

while y train and test won't be converted bcz they are already numbers(0&1 spam or ham)



two things are happening here
1. fitting the data in Vectorisor
2. converting all the data into feature vectors which is basiclll numerical values

*** we don't write extraction.fit_transform in x_ test bczwe don't wan our model to look at the test data

In [ ]:
print(X_train)

3075                  Don know. I did't msg him recently.
1787    Do you know why god created gap between your f...
1614                         Thnx dude. u guys out 2nite?
4304                                      Yup i'm free...
3266    44 7732584351, Do you want a New Nokia 3510i c...
                              ...                        
789     5 Free Top Polyphonic Tones call 087018728737,...
968     What do u want when i come back?.a beautiful n...
1667    Guess who spent all last night phasing in and ...
3321    Eh sorry leh... I din c ur msg. Not sad alread...
1688    Free Top ringtone -sub to weekly ringtone-get ...
Name: Message, Length: 4457, dtype: object


In [ ]:
print(X_train_features)

  (0, 5413)	0.6198254967574347
  (0, 4456)	0.4168658090846482
  (0, 2224)	0.413103377943378
  (0, 3811)	0.34780165336891333
  (0, 2329)	0.38783870336935383
  (1, 4080)	0.18880584110891163
  (1, 3185)	0.29694482957694585
  (1, 3325)	0.31610586766078863
  (1, 2957)	0.3398297002864083
  (1, 2746)	0.3398297002864083
  (1, 918)	0.22871581159877646
  (1, 1839)	0.2784903590561455
  (1, 2758)	0.3226407885943799
  (1, 2956)	0.33036995955537024
  (1, 1991)	0.33036995955537024
  (1, 3046)	0.2503712792613518
  (1, 3811)	0.17419952275504033
  (2, 407)	0.509272536051008
  (2, 3156)	0.4107239318312698
  (2, 2404)	0.45287711070606745
  (2, 6601)	0.6056811524587518
  (3, 2870)	0.5864269879324768
  (3, 7414)	0.8100020912469564
  (4, 50)	0.23633754072626942
  (4, 5497)	0.15743785051118356
  :	:
  (4454, 4602)	0.2669765732445391
  (4454, 3142)	0.32014451677763156
  (4455, 2247)	0.37052851863170466
  (4455, 2469)	0.35441545511837946
  (4455, 5646)	0.33545678464631296
  (4455, 6810)	0.29731757715898277
  (4

# Training the ML model

Logistic Model

In [ ]:
#loading the LR model in model variable
model = LogisticRegression()

In [ ]:
#training the LR model with the training data
#X_train_features ---> the feature vectors of the x tran i.e., msgs
#Y_train ---> corresponding 0, 1 value 
model.fit(X_train_features, Y_train)

LogisticRegression()

now the model is trained and it we will give and mail to it it will be able to categorise it into spam or ham mail but first we have to evaluate it

# Evaluating the trained model

In [ ]:
#prediction on training data only giving the X_train_features and 
#asking the model to predict the Y_ train value i.e., is the mail spam or ham (0 or 1)

prediction_on_training_data = model.predict(X_train_features) #corresponding Y_train is stored in prediction_on_training_data

accuracy_on_training_data = accuracy_score(Y_train, prediction_on_training_data)  #accuracy_score(real value, predicted value)


In [ ]:
print('Accuracy on training data = ', accuracy_on_training_data)

Accuracy on training data =  0.9670181736594121


that means accuracy is 96.7% 

we can say if the accuracy score is more than 75/5 than it is a good model and if >95% then it is working really well

In [ ]:
#prediction on testing data only giving the X_test_features and 
#asking the model to predict the Y_ test value i.e., is the mail spam or ham (0 or 1)

prediction_on_testing_data = model.predict(X_test_features) #corresponding Y_test is stored in prediction_on_testing_data

accuracy_on_testing_data = accuracy_score(Y_test, prediction_on_testing_data)  #accuracy_score(real value, predicted value)


In [ ]:
print('Accuracy on testing data = ', accuracy_on_testing_data)

Accuracy on testing data =  0.9659192825112107


in some cases the model may overfit(performs well on training data but not on testing data) that is why we tested it on both training and testing data

# Building a predictive system

In [ ]:
input_mail = ["I've been searching for the right words to thank you for this breather. I promise i wont take your help for granted and will fulfil my promise. You have been wonderful and a blessing at all times."]
#converting text to feature vectors
input_data_features = feature_extraction.transform(input_mail)

#making predictions
predictions = model.predict(input_data_features)
print(predictions)

if (predictions[0]==1):
  print('Ham mail')
else:
  print('Spam mail!!!')


[1]
Ham mail


predictions comes in a list

# Saving the trained model

In [ ]:
import pickle

In [ ]:
filename = 'trained_model.sav'
pickle.dump(model, open(filename, 'wb'))

In [ ]:
#loading the saved model
loaded_model = pickle.load(open('trained_model.sav', 'rb'))

In [ ]:
input_mail = ["I've been searching for the right words to thank you for this breather. I promise i wont take your help for granted and will fulfil my promise. You have been wonderful and a blessing at all times."]
#converting text to feature vectors
input_data_features = feature_extraction.transform(input_mail)

#making predictions
predictions = loaded_model.predict(input_data_features)
print(predictions)

if (predictions[0]==1):
  print('This mail is a Ham mail')
else:
  print('This mail is a Spam mail!!!')


[1]
This mail is a Ham mail
